### PR TITLE
Diagnostics: Add check for selected Kubernetes context.

### PR DESCRIPTION
**Pull request by mook-as**
It may be useful to review per-commit to reduce the amount of change to look at at once.

### CODE REVIEW

File: src/main/diagnostics/diagnostics.ts [hunk 1]
```diff
@@ -1,13 +1,18 @@
+import Logging from '@/utils/logging';
+
+const console = Logging.diagnostics;
+
 export enum DiagnosticsCategory {
-  Utilities = 'Utilities',
+  Kubernetse = 'Kubernetes',
```
Comment (ericpromislow) [inline src/main/diagnostics/diagnostics.ts:6]: typo in enum name

File: src/main/mainEvents.ts [hunk 2]
```diff
@@ -13,75 +13,140 @@ interface MainEventNames {
...
+   * Emitted as a reply to 'api-get-credentials'; the credentials can be used
+   * via HTTP basic auth on localhost.
+   *
+   * @note These credentials are meant for the UI; using them may require user
+   * interaction.
+   */
+  'api-credentials'(credentials: { user: string, password: string, port: number }): void;
 }
 
+type HandlerParams<eventName extends keyof MainEventNames> =
```
Comment (ericpromislow) [inline src/main/mainEvents.ts:84]: These three type def'ns need a comment explaining that they're only interested in events that return `void`, with an explanation of why.

File: src/main/mainEvents.ts [hunk 3]
```diff
@@ -13,75 +13,140 @@ interface MainEventNames {
...
+  void extends ReturnType<MainEventNames[eventName]> ? never :
+  (...args: HandlerParams<eventName>) => Promise<HandlerReturn<eventName>>;
+
 interface MainEvents extends EventEmitter {
   emit<eventName extends keyof MainEventNames>(
-    event: eventName,
+    event: void extends ReturnType<MainEventNames[eventName]> ? eventName : never,
     ...args: Parameters<MainEventNames[eventName]>
   ): boolean;
-  /* @deprecated */
```
Comment (ericpromislow) [inline src/main/mainEvents.ts:75]: Catching this now -- if `@deprecated` is a well-understood decorator, there should be a link to its doc. I don't see it at https://www.typescriptlang.org/docs/handbook/decorators.html and don't think I should have to search more deeply.

File: src/main/mainEvents.ts [hunk 4]
```diff
@@ -13,75 +13,140 @@ interface MainEventNames {
...
+   *
+   * @note These credentials are meant for the UI; using them may require user
+   * interaction.
+   */
+  'api-credentials'(credentials: { user: string, password: string, port: number }): void;
 }
 
+type HandlerParams<eventName extends keyof MainEventNames> =
+  void extends ReturnType<MainEventNames[eventName]> ? never :
+  Parameters<MainEventNames[eventName]>;
```
Comment (ericpromislow) [inline src/main/mainEvents.ts:86]: What does `Parameters<MainEventNames[eventName]>` resolve to?

File: src/main/mainEvents.ts [hunk 5]
```diff
@@ -13,75 +13,140 @@ interface MainEventNames {
...
+    handler: HandlerType<eventName>
+  ): void;
+}
+
+class MainEventsImpl extends EventEmitter implements MainEvents {
+  handlers: {
+    [eventName in keyof MainEventNames]?: HandlerType<eventName> | undefined;
+  } = {};
+
+  async invoke<eventName extends keyof MainEventNames>(
```
Comment (ericpromislow) [inline src/main/mainEvents.ts:131]: Another possibly dumb question -- what do these `invoke` and `handle` methods replace? Because I don't see anything that's been removed, but this code affects existing event names, not just the new diagnostic-related ones.

**Review by ericpromislow — CHANGES_REQUESTED** [review]
The code is working, but it needs some explanation, particularly in the details of how events are handled going forward.

File: src/main/diagnostics/diagnostics.ts [hunk 6]
```diff
@@ -87,17 +92,15 @@ export class DiagnosticsManager {
         import('./connectedToInternet'),
         import('./dockerCliSymlinks'),
         import('./rdBinInShell'),
+        import('./kubeContext'),
       ])).map(obj => obj.default);
-      const checkers = (await Promise.all(imports)).flat();
```
Comment (ericpromislow) [inline src/main/diagnostics/diagnostics.ts:91]: Not part of this PR, but I would rename `checkers` in `src/main/diagnostics/dockerCliSymlinks.ts` to something more specific, like `checkDockerCLIPlugins`.

File: src/main/mainEvents.ts (hunk 4 above)
Comment (mook-as) [reply src/main/mainEvents.ts:86]: I will add comments; to clarify, though, this gives you the parameter types of the method (because `eventName` names some method).  [`Parameters<T>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype) is a TypeScript builtin.

File: src/main/diagnostics/diagnostics.ts [hunk 7]
```diff
@@ -1,13 +1,18 @@
...
   Networking = 'Networking',
+  Utilities = 'Utilities',
 }
 
 /**
  * DiagnosticsCheckerResult is the result for running a given diagnostics
  * checker.
  */
-type DiagnosticsCheckerResult = {
+export type DiagnosticsCheckerResult = {
```
Comment (ericpromislow) [inline src/main/diagnostics/diagnostics.ts:15]: From an earlier commit, but this type has a mixture of single-star and jsdoc double-star comments. My IDE highlights the `/**...*/` ones, (and greys out the single-star ones), so maybe go with double-star.

**Review by ericpromislow — APPROVED** [review]
LGTM
